src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -15286,13 +15287,294 @@ public Vector512<T> Invoke(Vector512<T> x)
             }
         }
 
-        /// <summary>T.Round</summary>
-        internal readonly struct RoundOperator<T>(int digits, MidpointRounding mode) : IStatefulUnaryOperator<T> where T : IFloatingPoint<T>
+        /// <summary>T.Round(x)</summary>
+        internal readonly struct RoundToEvenOperator<T> : IUnaryOperator<T, T> where T : IFloatingPoint<T>
+        {
+            public static bool Vectorizable =>
+                (typeof(T) == typeof(float) || typeof(T) == typeof(double)) &&
+                (Avx2.IsSupported | AdvSimd.IsSupported);
+
+            public static T Invoke(T x) => T.Round(x);
+
+            private const uint Single_BiasedExponentMask = 0x7F80_0000;
+            private const int Single_BiasedExponentShift = 23;
+            private const uint Single_ShiftedBiasedExponentMask = (byte)(Single_BiasedExponentMask >> Single_BiasedExponentShift);
+            private const uint Single_TrailingSignificandMask = 0x007F_FFFF;
+
+            private const ulong Double_BiasedExponentMask = 0x7FF0_0000_0000_0000;
+            private const int Double_BiasedExponentShift = 52;
+            private const ulong Double_ShiftedBiasedExponentMask = (ushort)(Double_BiasedExponentMask >> Double_BiasedExponentShift);
+            private const ulong Double_TrailingSignificandMask = 0x000F_FFFF_FFFF_FFFF;
+
+            public static Vector128<T> Invoke(Vector128<T> x)
```

Comment:
Our current implementation is super inefficient, we should really just update it to do this like AMD's aocl-libm: https://github.com/amd/aocl-libm-ose/blob/master/src/optimized/nearbyint.c (double) and https://github.com/amd/aocl-libm-ose/blob/master/src/optimized/nearbyintf.c (float)

The logic, when ignoring the IEEE 754 exception handling we don't need, is basically:
```csharp
if (float.Abs(x) > 8388608.0f)
{
    return x;
}

float temp = float.CopySign(8388608.0f, x);
return float.CopySign((x + result) - result, x);
```

For `double`, the special boundary value is `4503599627370496.0` instead of `8388608.0f` -- these are notably just 2^52 and 2^23, which is the number of explicit mantissa bits each type carries.

Reply:
(This is meant to imply updating our scalar algorithm as well, it can be completely branch free and way simpler by doing this)

Reply:
-- The logic basically relies on the fact that IEEE 754 does the computation as if to infinite precision and unbounded range and then rounds to the nearest representable result. Since the default rounding mode is `to nearest, ties to even` already (and .NET doesn't support changing this) we can simplify the computation as above.

Anything that is "already correct" (inf, nan, or can't have a fractional portion) doesn't need computation. Otherwise, we push it above the boundary at which a fraction can't be represented.

The infinitely correct result then rounds to the nearest value and since every integer from `2^23` to `2^24` (or `2^52` to `2^53` for `double`) can be exactly represented, we get an infinitely precise integer that is `2^23` higher than the correct answer.

We do the copy sign to ensure negatives subtract and positives add, we remove the value to get the infinitely precise answer. The final copysign then exists to ensure that `-0.0` vs `+0.0` is correctly handled.

Reply:
Thanks. That algorithm is certainly easier.